fennel/datasets/test_schema_validator.py

```diff
@@ -1061,3 +1061,212 @@ def create_dataset(cls, activity: Dataset):
         str(e.value)
         == "Field `userid` is a key or timestamp field in schema of assign node input '[Dataset:RatingActivity]'. Value fields are: ['rating', 'movie']"
     )
+
```

Comment:
Very good tests. Suggestion to add couple more (if not already added):

1. It's valid to use `window` without groupby provided its operand already has some keys
2. It's an error if the output dataset doesn't have field called `window` (if not overriden) and the overriden field otherwise
3. The type of the window field must be `WindowStruct`
4. It's valid to add operators after window (e.g. filter, transform)
5. It's valid to compute aggregate after window (e.g. avg window duration, which one of the customer is asking for)

Reply:
Will add tests 2, 3, 4, 5 about 1 I had a doubt I raised in previous comment also.

Reply:
+1 A few more tests where Window is an intermediate dataset and there are further downstream datasets

Reply:
Also a check if any key is optional

Reply:
For key ti be optional, I think schema validator for groupBy will handle that, I will write tests for the same.

Reply:
added.